modules/auxiliary/scanner/msmq/cve_2023_21554_queuejumper.rb

```diff
@@ -0,0 +1,259 @@
+##
+# This module requires Metasploit: https://metasploit.com/download
+# Current source: https://github.com/rapid7/metasploit-framework
+##
+
+class MetasploitModule < Msf::Auxiliary
+  include Msf::Exploit::Remote::Tcp
+  include Msf::Auxiliary::Scanner
+  include Msf::Auxiliary::Report
+
+  def initialize(info = {})
+    super(
+      update_info(
+        info,
+        'Name' => 'CVE-2023-21554 - QueueJumper - MSMQ RCE Check',
+        'Description' => %q{
+          This module checks the provided hosts for the CVE-2023-21554 vulnerability by sending
+          a MSMQ message with an altered DataLength field within the SRMPEnvelopeHeader that
+          overflows the given buffer. On patched systems, the error is catched and no response
+          is sent back. On vulnerable systems, the integer wraps around and depending on the length
+          could cause an out-of-bounds write. In the context of this module a response is sent back,
+          which indicates that the system is vulnerable.
+        },
+        'Author' => [
+          'Wayne Low and Haifei Li', # Vulnerability discovery
+          'Bastian Kanbach <bastian.kanbach@securesystems.de>' # Metasploit Module, @__bka__
+        ],
+        'References' => [
+          [ 'CVE', '2023-21554' ],
+          [ 'URL', 'https://msrc.microsoft.com/update-guide/vulnerability/CVE-2023-21554' ],
+          [ 'URL', 'https://securityintelligence.com/posts/msmq-queuejumper-rce-vulnerability-technical-analysis/' ]
+        ],
+        'DisclosureDate' => '2023-04-11',
+        'License' => MSF_LICENSE,
+        'Notes' => {
+          'Stability' => [ CRASH_SAFE ],
+          'SideEffects' => [IOC_IN_LOGS],
+          'Reliability' => [REPEATABLE_SESSION],
+          'AKA' => ['QueueJumper']
+        }
+      )
+    )
+    register_options([
+      Opt::RPORT(1801)
+    ])
+  end
+
+  def message_normal
+    "\x10\x00\x03\x10\x4c\x49\x4f\x52\x64\x09\x00\x00\x63\x76\x09\x6c" \
```

Comment:
Maybe add some comment explaining where this huge binary blob is coming from and what it contains?

Reply:
Very good point. I did some changes where I splitted the huge message struct up into its different headers. To some extent these are still huge, however I also added some documentation for each header, detailing what fields they use and what the purpose of the headers are.

Reply:
Anything else you would suggest to add, before merge? Thanks for the review

Reply:
I think this thread is resolved now by the swap to bindata, thanks!